ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://mesolite.cluster.engflow.com/invocations/default/98da77f1-4c35-4b0f-bf6f-04c4bc22d1c1?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-25.3 @ [0a40cdd22f3a1ff489467c1b3208b7e7739f6b0d](https://github.com/cockroachdb/cockroach/commits/0a40cdd22f3a1ff489467c1b3208b7e7739f6b0d):

Fatal error:

```
panic: kvFeed: unexpected 10240 leftover bytes

```
Stack: 

```
goroutine 27862760 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x8715038, 0xc015adb680}, 0xc0061f0000, {0x7820f77?, 0xc004c33790?}, {0xc004c337d0?, 0x1?, 0x4c33748?})
	pkg/util/log/logcrash/crash_reporting.go:434 +0x1e7
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc00e8e8d10, {0x8715038, 0xc015adb680}, 0x1)
	pkg/util/mon/bytes_usage.go:796 +0x36a
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	pkg/util/mon/bytes_usage.go:722
runtime.Goexit()
	GOROOT/src/runtime/panic.go:635 +0x5e
testing.(*common).FailNow(0xc00a60f6c0)
	GOROOT/src/testing/testing.go:1006 +0x4a
github.com/stretchr/testify/require.Error({0x86ce9d8, 0xc00a60f6c0}, {0x0, 0x0}, {0x0, 0x0, 0x0})
	external/com_github_stretchr_testify/require/require.go:293 +0xca
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.runWithAndWithoutRegression141453.func3.1.1.4({0x0?, 0x0?})
	pkg/ccl/changefeedccl/helpers_test.go:1941 +0x45
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent.(*blockingBuffer).Drain.func1()
	pkg/ccl/changefeedccl/kvevent/blocking_buffer.go:331 +0x29
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent.(*blockingBuffer).Drain(0xc00e8e8dc0?, {0x8715038?, 0xc015adb680?})
	pkg/ccl/changefeedccl/kvevent/blocking_buffer.go:344 +0x108
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.Run({_, _}, {0xc0061f0000, 0xc011f68e60, {{0xc0167b9710, 0xc0167b9728, {0xa}}, {0xc0167b9710}, 0x0}, 0xc00365fb20, ...})
	pkg/ccl/changefeedccl/kvfeed/kv_feed.go:176 +0xcf2
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).startKVFeed.func1({0x8715038?, 0xc015adb680?})
	pkg/ccl/changefeedccl/changefeed_processors.go:481 +0xf8
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x8715038, 0xc015adb680})
	pkg/util/stop/stopper.go:438 +0x8a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 27862579
	pkg/util/stop/stopper.go:436 +0x137
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed4189034041
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/regression_141453=true
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/regression_141453=true/sinkless
    helpers_test.go:1238: making server as secondary tenant
    helpers_test.go:1317: making sinkless feed factory
    helpers_test.go:1382: pgURL sinkless SinklessFeedUser
    helpers_test.go:1382: pgURL sinkless root
    helpers_test.go:1053: did not force enriched envelope for CREATE CHANGEFEED FOR TABLE mytable because *changefeedccl.sinklessFeedFactory is not supported
    testfeed_test.go:294: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    helpers_test.go:1941: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:1941
        	            				pkg/ccl/changefeedccl/kvevent/blocking_buffer.go:331
        	            				pkg/ccl/changefeedccl/kvevent/blocking_buffer.go:344
        	            				pkg/ccl/changefeedccl/kvfeed/kv_feed.go:176
        	            				pkg/ccl/changefeedccl/changefeed_processors.go:481
        	            				pkg/util/stop/stopper.go:438
        	            				src/runtime/asm_amd64.s:1700
        	Error:      	An error is expected but got nil.
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/regression_141453=true/sinkless
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57123

--- a/pkg/ccl/changefeedccl/kvfeed/kv_feed.go
+++ b/pkg/ccl/changefeedccl/kvfeed/kv_feed.go
@@ -104,7 +104,30 @@ type Config struct {
 
 // Run will run the kvfeed. The feed runs synchronously and returns an
 // error when it finishes.
-func Run(ctx context.Context, cfg Config) error {
+func Run(ctx context.Context, cfg Config) (retErr error) {
+	// The writer must always be closed to release resources and signal to the
+	// consumer (changeAggregator) that no more writes are expected, allowing it
+	// to transition to draining state. The defer ensures cleanup even if Drain
+	// panics, while the success path explicitly closes before waiting for the
+	// changeAggregator to finish.
+	//
+	// Closed tracks whether we've already closed the writer, preventing the
+	// defer from closing twice.
+	closed := false
+
+	defer func() {
+		if closed {
+			return
+		}
+		closeReason := retErr
+		if closeReason == nil {
+			// A nil retErr indicates a panic occurred.
+			closeReason = errors.New("kvfeed terminated abnormally")
+		}
+		if closeErr := cfg.Writer.CloseWithReason(ctx, closeReason); closeErr != nil {
+			retErr = errors.CombineErrors(retErr, closeErr)
+		}
+	}()
 
 	var sc kvScanner
 	{
@@ -153,8 +176,8 @@ func Run(ctx context.Context, cfg Config) error {
 	if !isChangefeedCompleted && !errors.As(err, &scErr) {
 		log.Changefeed.Errorf(ctx, "stopping kv feed due to error: %s", err)
 		// Regardless of whether we exited KV feed with or without an error, that error
-		// is not a schema change; so, close the writer and return.
-		return errors.CombineErrors(err, f.writer.CloseWithReason(ctx, err))
+		// is not a schema change; so, return and let the defer handle cleanup.
+		return err
 	}
 
 	if isChangefeedCompleted {
@@ -165,14 +188,14 @@ func Run(ctx context.Context, cfg Config) error {
 
 	// Drain the writer before we close it so that all events emitted prior to schema change
 	// or changefeed completion boundary are consumed by the change aggregator.
-	// Regardless of whether drain succeeds, we must also close the buffer to release
-	// any resources, and to let the consumer (changeAggregator) know that no more writes
-	// are expected so that it can transition to a draining state.
-	if err := f.writer.Drain(ctx); err != nil {
-		err := errors.Wrap(err, "failed to drain kv feed writer")
-		return errors.CombineErrors(err, f.writer.CloseWithReason(ctx, err))
+	if drainErr := f.writer.Drain(ctx); drainErr != nil {
+		return errors.Wrap(drainErr, "failed to drain kv feed writer")
 	}
-	if err := f.writer.CloseWithReason(ctx, kvevent.ErrNormalRestartReason); err != nil {
+
+	// Close the writer to signal completion, then wait for the
+	// changeAggregator to finish consuming all buffered events.
+	closed = true
+	if err := cfg.Writer.CloseWithReason(ctx, kvevent.ErrNormalRestartReason); err != nil {
 		return err
 	}
 

